build: introduce configure --shared

stefanmb opened the pull request:
<!--
Thank you for your pull request. Please review below requirements and walk
through the checklist. You can 'tick' a box by using the letter "x": [x].

Run the test suite with: `make -j4 test` on UNIX or `vcbuild test nosign` on
Windows.

If this aims to fix a regression or you’re adding a feature, make sure you also
write a test. If possible, include a benchmark that quantifies your changes.

Finally, read through our contributors guide and make adjustments as necessary:
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- remove lines that do not apply to you -->

- [x] tests and code linting passes
- [x] documentation is changed or added
- [x] the commit message follows commit guidelines


##### Affected core subsystem(s)
<!-- provide affected core subsystem(s) (like doc, cluster, crypto, etc) -->

build, configure

##### Description of change
<!-- provide a description of the change below this comment -->

This PR is the continuation of https://github.com/nodejs/node/pull/5918 where the originator (@indutny) agreed that I can take over this task in order to land the changes as soon as possible.

The PR consists of two commits, the original one from https://github.com/nodejs/node/pull/5918 which adds "[a] configure flag for building shared library that could be used to embed Node.js in some application (like Electron)" and a second commit that modifies the following aspects:

- Use the "--shared" configure flag to build a shared library.
- Use the "--no-bundled-v8" flag to exclude V8 headers from the deps/ folder, this option is useful to Electron which provides its own copy of V8.
- Use the "--no-v8-platform" flag to avoid initializing the V8 platform inside Node (useful for some embedders).
- Insure that "./configure --shared && make" works and produces a useable libnode.so for anyone (i.e. decouple --shared from Electron's build system).

I intend to land both commits in the PR at the same time (the original one from @indutny and my changes on top of it) to maintain a clear history/ownership of the work. If anyone feels strongly about this approach either way, please let me know.

All these options are provided as-is for embedders are not officially supported at this time, as indicated in the documentation.

**Future changes:**

This PR is concerned with Linux only - once this PR is landed a separate PR will be made to resolve issues related to building a DLL on Windows.

stefanmb commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/2808/

TheAlphaNerd commented:
@stefanmb it looks like this needs a rebase. Let me know if there is anything I can help with on this

rvagg commented:
@nodejs/build we should probably reach out to distro packagers to get feedback on this, if it's not useful to them all then it's probably not worth it as it is. Who do we have? @kapouer is the only one I can think of off the top of my head.

kapouer commented:
Yes, why not. I'll try to use that option later today and report back.

kapouer commented:
When building a shared library, it is a common and good practice to give it a soname.
v8 contains an example of how to do so with gyp, just grep `soname` there.
I think the right soname for node is `libnode.so.${process.versions.modules}`.

bnoordhuis commented:
/cc @sgallagher - see https://github.com/nodejs/node/pull/6994#issuecomment-222590407.

sgallagher commented:
Passing the buck over to @thughes

mhdawson commented:
We are interested in it because we have internal product teams that what to include Node.js in a product without having a separate process.  So I don't see that it should be tied to the distro packages.  If they say it would be useful to have a shared library but this does not address the requirement then we need that input and need to adjust.  If on the other hand they just don't need a shared library I don't think that invalidates its usefulness as we have examples like electron and our internal team which do want it as a shared library.

kapouer commented:
@mhdawson on the debian side i'm pretty sure it's going to be useful, one way or another.

mhdawson commented:
@kapouer good to hear.

stefanmb commented:
@TheAlphaNerd Rebased onto master, thanks.

@rvagg @mhdawson In addition to Michael's comments, I think the need for this feature was strongly expressed in https://github.com/nodejs/node/pull/5918, https://github.com/nodejs/node-v0.x-archive/pull/7336, and https://github.com/nodejs/roadmap/issues/9. It is my intention to get this feature working on all platforms starting with Linux in this PR.

@kapouer Thank you for the soname suggestion, I've implemented it (and fixed the install.py script which I'd previously broken due to some bad merging, sorry!). It required a few hacks to get the information in the right place but the end product will now be "libnode.so.${process.versions.modules}" as you indicated. This may be a bit counter-intuitive to regular users since the module version does not map to the Node release version in an obvious way, but I believe it makes sense to provide this naming convention for embedders.

New CI: https://ci.nodejs.org/job/node-test-pull-request/2880/

stefanmb commented:
One unrelated failure (socket timeout) on ARM: https://ci.nodejs.org/job/node-test-binary-arm/RUN_SUBSET=0,label=pi1-raspbian-wheezy/2368/console

jbergstroem commented:
@kapouer do you know of any packages that would use this?

kapouer commented:
@jbergstroem besides addons (it's more natural to link them to libnode) no, not yet. Electron, eventually...

jbergstroem commented:
@kapouer ok, thanks. How about libv8 bindings? Would any of those deps overflow/consider switching?

kapouer commented:
Could you be a little more specific ? I don't understand the question.

jbergstroem commented:
@kapouer sorry; was just wondering if packages that builds against a shared v8 would/could be interested in this.

kapouer commented:
@jbergstroem i'm not sure it would be a good idea to link to libnode for a software that just needs v8. Maybe one day libnode will expose only nan-like symbols ?

stefanmb commented:
Based on some comments from @mhdawson I've made the following changes:

* Rebase again onto master.
* Fixed `NODE_CTOR_PREFIX` to be guarded under the correct `NODE_SHARED_MODE` define. This define will remove `static` from the DSO constructors for the modules when in shared library mode. It's not strictly needed to do so in order to have the constructor symbols exported, but Electron uses them so I've kept compatibility, see https://github.com/nodejs/node-v0.x-archive/pull/7336#issuecomment-47450600 for details.
* Remove superfluous dependencies in node.gyp.
* Fixed some rules in node.gyp to check for `no_bundled_v8` instead of `node_shared`

CI: https://ci.nodejs.org/job/node-test-pull-request/2906/

mhdawson commented:
LGTM with one question I'd like to have confirmation for.  I assume that post-mortem and vtune support is disabled for node_no_bundled_v8 because we need the specific v8 version/code to support those which we don't have when that option is supported, right ?

stefanmb commented:
@mhdawson Those options both introduce dependencies from the bundled V8 folder:
`deps/v8/src/third_party/vtune/v8vtune.gyp:v8_vtune` and `deps/v8/tools/gyp/v8.gyp:postmortem-metadata` respectively. The `node_no_bundled_v8` option is meant to exclude dependencies on the bundled V8.

mhdawson commented:
@indutny can you take a look at this PR to confirm it covers the requirements your original PR was intended to address ?

jasnell commented:
LGTM with a rebase, green CI, and @indutny's ok :-)

stefanmb commented:
@jasnell Thanks, rebased. CI: https://ci.nodejs.org/job/node-test-pull-request/2979

indutny commented:
LGTM if CI is green.

thefourtheye commented on configure:
```diff
@@ -24,6 +24,10 @@ from gyp.common import GetFlavor
 import nodedownload
 
+# imports in tools/
+sys.path.insert(0, os.path.join(root_dir, 'tools', ))
```
extra comma after `tools`

thefourtheye commented on node.gyp:
```diff
@@ -215,6 +222,37 @@
+        }],
+        [ 'node_no_bundled_v8=="false"', {
+          'include_dirs': [
+            'deps/v8', # include/v8_platform.h
```
Is this comment necessary?

bnoordhuis replied:
It exists in the node.gyp in master too.

It makes sense to include it because I don't think it would be clear to a casual reader why deps/v8 rather than deps/v8/include should be on the include path.

It's because some headers `#include "include/v8_platform.h"` instead of `#include "v8_platfom.h"`.  That's fixed in V8 5.1 but we can't remove it until we upgrade.

thefourtheye replied:
oh okay then.

stefanmb commented:
@thefourtheye Thanks, I removed the trailing comma. One last CI: https://ci.nodejs.org/job/node-test-pull-request/2995/

stefanmb commented:
Unrelated failures in the last run, ran again: https://ci.nodejs.org/job/node-test-pull-request/3001/ and got these on CentOS:
* 52 - parallel/test-child-process-fork-dgram 
* 558 - parallel/test-https-connect-address-family  
* 943 - parallel/test-tls-connect-address-family  

I think it's safe to say they are unrelated, will merge later today.

bnoordhuis commented on node.gyp:
```diff
@@ -477,20 +516,14 @@
         }],
         [ 'OS=="freebsd" or OS=="linux"', {
-          'ldflags': [ '-Wl,-z,noexecstack',
+          'ldflags': [ '-Wl,-z,noexecstack,--allow-multiple-definition',
```
Why is `--allow-multiple-definition` necessary?

stefanmb replied:
Without it there are linking errors related to OpenSSL and V8, example:

`/out/Release/obj.target/deps/openssl/libopenssl.a(bio_ssl.o): In function 'BIO_f_ssl':
bio_ssl.c:(.text+0xb00): multiple definition of 'BIO_f_ssl'
/out/Release/libopenssl.a(bio_ssl.o):bio_ssl.c:(.text+0xb00): first defined here
/out/Release/obj.target/deps/openssl/libopenssl.a(bio_ssl.o): In function 'BIO_new_ssl':
bio_ssl.c:(.text+0xb10): multiple definition of 'BIO_new_ssl'
/out/Release/libopenssl.a(bio_ssl.o):bio_ssl.c:(.text+0xb10): first defined here
/out/Release/obj.target/deps/openssl/libopenssl.a(bio_ssl.o): In function 'BIO_new_ssl_connect':
bio_ssl.c:(.text+0xba0): multiple definition of 'BIO_new_ssl_connect'
/out/Release/libopenssl.a(bio_ssl.o):bio_ssl.c:(.text+0xba0): first defined here
/out/Release/obj.target/deps/openssl/libopenssl.a(bio_ssl.o): In function 'BIO_new_buffer_ssl_connect':
bio_ssl.c:(.text+0xc10): multiple definition of 'BIO_new_buffer_ssl_connect'
/out/Release/libopenssl.a(bio_ssl.o):bio_ssl.c:(.text+0xc10): first defined here
/out/Release/obj.target/deps/openssl/libopenssl.a(bio_ssl.o): In function 'BIO_ssl_copy_session_id':
bio_ssl.c:(.text+0xc90): multiple definition of 'BIO_ssl_copy_session_id'
`

bnoordhuis replied:
That sounds like (for example) a static library is getting linked in twice.  I'm inclined to say that `--allow-multiple-definition` is a workaround but not a proper fix.

bnoordhuis commented on src/node.cc:
```diff
@@ -183,7 +185,41 @@ static uv_async_t dispatch_debug_messages_async;
+  }
+#endif  // HAVE_INSPECTOR
+
+  v8::Platform* platform;
```
`platform_`

stefanmb replied:
Will fix.

bnoordhuis commented on src/node.cc:
```diff
@@ -183,7 +185,41 @@ static uv_async_t dispatch_debug_messages_async;
 static v8::Isolate* node_isolate;
-static v8::Platform* default_platform;
+
+static struct v8_plat {
```
You might as well drop the struct name if you're not using it anywhere.

bnoordhuis commented on src/node.h:
```diff
@@ -409,17 +409,23 @@ extern "C" NODE_EXTERN void node_module_register(void* mod);
+# define NODE_CTOR_PREFIX
+#else
+# define NODE_CTOR_PREFIX static
+#endif
```
One result of removing `static` (if I read the changes right), is that internal module constructors leak into the list of exported symbols.  Or is that mitigated by `-fvisibilty=hidden`?

stefanmb replied:
@bnoordhuis Correct, here is an example  (notice uppercase T for export): 

```
nm -a out/Release/lib.target/libnode.so.48 | grep _register_
...
0000000000622d90 T _register_async_wrap
0000000000622e10 T _register_buffer
0000000000622dd0 T _register_cares_wrap
0000000000622e30 T _register_config
0000000000622e50 T _register_contextify
0000000000623070 T _register_crypto
0000000000622e70 T _register_fs
0000000000622db0 T _register_fs_event_wrap
0000000000622e90 T _register_http_parser
0000000000622df0 T _register_js_stream
0000000000622eb0 T _register_os
0000000000622f30 T _register_pipe_wrap
0000000000623010 T _register_process_wrap
0000000000622f50 T _register_signal_wrap
0000000000622f70 T _register_spawn_sync
0000000000622f90 T _register_stream_wrap
0000000000622fb0 T _register_tcp_wrap
0000000000622fd0 T _register_timer_wrap
00000000006230d0 T _register_tls_wrap
0000000000622ff0 T _register_tty_wrap
0000000000623030 T _register_udp_wrap
0000000000622ed0 T _register_util
0000000000623050 T _register_uv
0000000000622ef0 T _register_v8
0000000000622f10 T _register_zlib
```

This change was in the original PR by Fedor. I believe he included it due to Electron, see here https://github.com/electron/node/commit/0828dfa193038209b7b337ba5479c3373f7cf87f and the uses here https://github.com/electron/electron/commit/69adff19fe610800de900fc9358a62a402769bac

bnoordhuis commented on tools/getmoduleversion.py:
```diff
@@ -0,0 +1,19 @@
+
+def get_version():
+  node_version_h = os.path.join(os.path.dirname(__file__), '..', 'src',
+    'node_version.h')
```
Style: arguments should line up.

bnoordhuis commented on tools/getmoduleversion.py:
```diff
@@ -0,0 +1,19 @@
+
+  f = open(node_version_h)
+
+  regex = '#define NODE_MODULE_VERSION [0-9]+'
```
You should anchor (`^`) this to the start of the line so it won't catch e.g. a comment.

bnoordhuis commented on tools/getmoduleversion.py:
```diff
@@ -0,0 +1,19 @@
+  
+  raise Exception("Could not find pattern matching '%s'" % regex)
+
+if __name__ == "__main__":
```
Can you use single quotes here?

bnoordhuis commented:
I'm not a fan of `node_no_foo` variable names.  Can I suggest you invert the logic and call them `node_use_foo`?

empyrical commented:
> This PR is concerned with Linux only - once this PR is landed a separate PR will be made to resolve issues related to building a DLL on Windows.

Just thought I'd give you a heads up that it works fine, built and linked dynamically in a simple c++ script, on Mac OS 10.11.5!

One thing I think this will need is a small addition to the node bootstrapper that can give embedders a way to programatically provide a string or something to be their main. There does seem to be one official way to make your own main, but you need to include a native module called `_third_party_main` at compile time. Unless I'm misunderstanding something just creating a node environment will pepper the JS globals with a `process` object among other things, but it seems that node will only put `require` into the global namespace if an appropriate main function is ran.

kapouer commented:
I knew it would be useful to us, debian users:
Being able to use Node.js and system-installed addons (if any) in a MultiArch setup requires linking to be done to a shared library (because MultiArch deals with library paths like `/usr/lib/x86_64-linux-gnu/` and debian policy require files put there to be shared libraries with sonames).
From [MultiArch wiki](https://wiki.debian.org/Multiarch/HOWTO):
> Multiarch lets you install library packages from multiple architectures on the same machine.

It would even be useful on current LTS...

stefanmb commented:
@bnoordhuis Thanks for the review, I've addressed the issues you raised. I've eliminated the use of `--allow-multiple-definitions`, I made the necessary style changes to the Python scripts, and inverted the logic of `node_no_foo` to be `node_use_foo`, and changed the naming inside the C struct. 

CI: https://ci.nodejs.org/job/node-test-pull-request/3070/

bnoordhuis commented on configure:
```diff
@@ -420,6 +424,26 @@ parser.add_option('--without-inspector',
+    help='do not initialize v8 platform during node.js startup. ' +
+         '(This mode is not officially supported for regular applications)')
+
+parser.add_option('--no-bundled-v8',
```
Can you call these `--without-<option>` for consistency?

bnoordhuis commented on configure:
```diff
@@ -810,6 +834,12 @@ def configure_node(o):
+  o['variables']['node_use_bundled_v8'] = b(not options.no_bundled_v8)
+  
+  if flavor == 'linux' and options.shared:
+    o['variables']['node_module_ver'] = getmoduleversion.get_version()
```
I'd be okay with exporting this unconditionally.  It's going to end up in `process.config.variables` where it might be useful (for example) for people building add-ons.  I'd call it `node_module_version` though.

bnoordhuis commented on node.gyp:
```diff
@@ -100,6 +103,13 @@
+        'node_target_type%': 'shared_library',
+      }, {
+        'node_target_type%': 'executable',
+      }],
```
This breaks `./configure --static`.

bnoordhuis commented on node.gyp:
```diff
@@ -308,7 +347,7 @@
                   },
                   'conditions': [
-                    ['OS in "linux freebsd"', {
+                    ['OS in "linux freebsd"' and 'node_shared=="false"', {
```
This looks wrong; too many single quotes.

bnoordhuis commented on node.gyp:
```diff
@@ -217,6 +224,37 @@
+          ],
+        }, {
+          'defines': [
+            'NODE_NO_V8_PLATFORM',
```
Can you call this `NODE_USE_V8_PLATFORM` and set it to 0 or 1 depending on whether `node_use_v8_platform` is true.

(If you make an int, you could shorten it to `'defines': [ 'NODE_USE_V8_PLATFORM=<(node_use_v8_platform)' ]`, although that would be inconsistent with the other node_ variables.)

bnoordhuis commented on tools/getmoduleversion.py:
```diff
@@ -0,0 +1,22 @@
+import os,re
```
One import per line, please (and in sorted order.)

bnoordhuis commented on tools/getmoduleversion.py:
```diff
@@ -0,0 +1,22 @@
+  raise Exception('Could not find pattern matching %s' % regex)
+
+if __name__ == '__main__':
+  print get_version()
```
If you add `from __future__ import print_function` at the top, then you can write `print(get_version())` here.  A little more future-proof.

bnoordhuis commented on tools/install.py:
```diff
@@ -107,9 +108,21 @@ def subdir_files(path, dest, action):
+      # GYP will output to lib.target, this is hardcoded in its source,
+      # see the _InstallableTargetInstallPath function.
+      output_prefix += 'lib.target/'
+      output_file = 'lib' + output_file + '.so.' + get_version()
```
This also runs when node_shared=="false", which seems wrong?

stefanmb replied:
Should be resolved now, thanks.

stefanmb commented:
@bnoordhuis I addressed most of the comments. WIth respect to `./configure --enable-static`, it works fine during my tests. It's a concern (and configure time option) that's orthogonal to `./configure --shared`.

CI: https://ci.nodejs.org/job/node-test-pull-request/3095/

bnoordhuis commented on node.gyp:
```diff
@@ -217,6 +224,42 @@
+        }],
+        [ 'node_use_v8_platform=="true"', {
+          'defines': [
+            'NODE_NO_V8_PLATFORM=0',
```
Can you call the defin NODE_USE_V8_PLATFORM?

bnoordhuis commented:
Can you add a small test that checks that `process.config.node_module_version` is a number > 0?

stefanmb commented:
@bnoordhuis Inverted the logic for the define, made `node_module_version` an integer and added a small test in `test/parallel/test-module-version.js`.

CI: https://ci.nodejs.org/job/node-test-pull-request/3106/

stefanmb commented:
CI showed some unrelated failures (timeout, file system issues on Jenkins).

bnoordhuis commented on src/node.cc:
```diff
@@ -218,7 +209,17 @@ static struct {
+  void Dispose() {}
+#if HAVE_INSPECTOR
+  void StartInspector(Environment *env, int port, bool wait) {
+    env->ThrowError("Node compiled with NODE_NO_V8_PLATFORM");
```
s/NODE_NO_V8_PLATFORM/NODE_USE_V8_PLATFORM=0/

stefanmb commented:
@bnoordhuis Oops, missed that line when I reshuffled things. The error message should be updated now, thanks again for the reviews.

bnoordhuis commented:
LGTM

stefanmb commented:
Thank you all for the help! Landed as https://github.com/nodejs/node/commit/410296c37acb2ad5d939bd4633b2c3fbe3782084.

joshgav commented:
Thank you @stefanmb! Great stuff!

Per your original post, is there still a plan for a Windows DLL?

/cc @orangemocha 

> Future changes:

> This PR is concerned with Linux only - once this PR is landed a separate PR will be made to resolve issues related to building a DLL on Windows.

stefanmb commented:
@joshgav Yes, the Windows changes are coming. The main problem is making sure that multiple copies of the CRT are not included, which may mean some changes to the V8 gyp files (basically OpenSSL, V8, and Node have to be compiled using Multi-threaded DLL (/MD) with the same version of MSVC to avoid a whole slew of problems).

mhdawson commented:
@joshgav, just to add to @stefanmb's comments, this is just the first step.  Adding support for the rest of the platforms (windows, aix) and adding testing to the CI to name a few.  Windows may take a bit longer as we'll have to get changes landed in V8 master before we can float them in Node.

TheAlphaNerd commented:
@stefanmb @mhdawson lts?

stefanmb commented:
@TheAlphaNerd I would say no since it involves a floating patch to V8? I'm not sure!

TheAlphaNerd commented:
@stefanmb we are meeting with V8 on wednesday to discuss our own release lines for LTS so we will no longer have to float patches. Let's see how that turns out before making a decision on this

bnoordhuis commented:
If this PR does get back-ported to v4.x, it should include the changes from e.g. https://github.com/nodejs/node/pull/7538 that have been dropped due to merge conflicts.

RReverser commented:
I'm likely just missing it, but do we have any docs on the exported C / C++ APIs when using node as a library or is source code the only source of truth?

mhdawson commented:
@RReverser, I believe the code is the only source of truth as you say.  We are still working on adding support for other platforms (windows, OSX, AIX) and then adding testing so will keep adding some doc in mind as well.

TheAlphaNerd commented:
The working group discussed this pull request and have decided to land it in a future v4.x release

@mhdawson would you be able to take the lead on backporting?

dominictarr commented:
My understanding is also that this would be essential for running on a unrooted andoid. you get one process, which must be java, but you can link a .so which could be `libnode.so`. That is how jxcore (RIP) did it.

TheAlphaNerd commented:
@dominictarr you should find this commit now in the latest v4, v6, and v7

I believe there are still other edges regarding android... let me know if you try and get it working
